flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java

```diff
@@ -549,6 +551,157 @@ public void testConsistentAvailability() throws Exception {
+
+        testUseAllBuffersForSingleChannel(12, 10, maxOverdraftBuffers);
+        testUseAllBuffersForMultiChannel(12, 10, maxOverdraftBuffers);
+    }
+
+    @Test
+    public void testWithOverdraftBuffer() throws Exception {
+        localBufferPool.lazyDestroy();
+        int maxOverdraftBuffers = 2;
+
+        testUseAllBuffersForSingleChannel(4, 3, maxOverdraftBuffers);
+        testUseAllBuffersForMultiChannel(4, 3, maxOverdraftBuffers);
+
+        testUseAllBuffersForSingleChannel(8, 5, maxOverdraftBuffers);
+        testUseAllBuffersForMultiChannel(8, 5, maxOverdraftBuffers);
+
+        testUseAllBuffersForSingleChannel(12, 10, maxOverdraftBuffers);
+        testUseAllBuffersForMultiChannel(12, 10, maxOverdraftBuffers);
+    }
+
+    private void testUseAllBuffersForSingleChannel(
+            int poolSize, int maxBuffersPerChannel, int maxOverdraftBuffers) throws Exception {
+        int numberOfSubpartitions = 2;
+        checkArgument(maxBuffersPerChannel > poolSize / numberOfSubpartitions);
+        LocalBufferPool bufferPool =
+                new LocalBufferPool(
+                        networkBufferPool,
+                        1,
+                        Integer.MAX_VALUE,
+                        numberOfSubpartitions,
+                        maxBuffersPerChannel,
+                        maxOverdraftBuffers);
+        bufferPool.setNumBuffers(poolSize);
+
+        // Request all buffers inside the buffer pool
+        AutoCloseableRegistry channelCloseableRegistry = new AutoCloseableRegistry();
+        for (int i = 0; i < maxBuffersPerChannel; i++) {
+            assertRequestedOverdraftBufferAndIsAvailable(bufferPool, 0, true);
+            BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(0);
+            assertNotNull(bufferBuilder);
+            channelCloseableRegistry.registerCloseable(bufferBuilder);
+        }
+
+        AutoCloseableRegistry poolCloseableRegistry = new AutoCloseableRegistry();
+        for (int i = maxBuffersPerChannel; i < poolSize; i++) {
+            assertRequestedOverdraftBufferAndIsAvailable(bufferPool, 0, false);
+            BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(0);
+            assertNotNull(bufferBuilder);
+            poolCloseableRegistry.registerCloseable(bufferBuilder);
+        }
+
+        // request overdraft buffer
+        AutoCloseableRegistry overdraftCloseableRegistry = new AutoCloseableRegistry();
+        assertRequestedOverdraftBufferAndIsAvailable(bufferPool, 0, false);
+        for (int i = 0; i < maxOverdraftBuffers; i++) {
+            BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(0);
+            assertNotNull(bufferBuilder);
+            overdraftCloseableRegistry.registerCloseable(bufferBuilder);
+            assertRequestedOverdraftBufferAndIsAvailable(bufferPool, i + 1, false);
+        }
+
+        BufferBuilder bufferBuilder = bufferPool.requestBufferBuilder(1);
```

Comment:
I think that it makes sense to check that `requestBufferBuilder` for `0` channel returns NULL value as well